### ISSUE TITLE
Inconsistent results when using tf_upgrade_v2 script

### ISSUE THREAD

**Issue by tatianacv** [tensorflow/tensorflow#45585]
**System information**

- Run in Google Collab Notebook with macOS Big Sur
- There are two files run: 
    - a version running code with Tensorflow 2
    - a version running code with Tensorflow 1.x
- Python 3.6.9

**Current behavior**

We wanted to see if the result of the migration script from TF 1.x to TF 2 gives the same output. To evaluate this, we took code from Tensorflow 1.x taken from this tutorial (https://medium.com/@udolf15/mnist-digits-classification-with-tensorflow-7f7dcda0fc1e) and ran it with the migration script and got a compat version of the code. I set the graph-level and operation-level seeds to be able to have reproducible outputs on both the compat(TF2 using migration script) and original version (TF 1.x). 

The results from running the Original TF 1.x version are: 
```
(train_loss, train_accuracy)
0.8028196 0.91954
0.23598278 0.9466
0.1635617 0.96346
0.12261735 0.97066
0.09810413 0.97426
0.07586458 0.97328
0.057922967 0.98538
0.0451939 0.98892
0.036826767 0.99028
0.032206446 0.98698
0.027886271 0.99306
0.023077298 0.99478
0.016734073 0.99372
0.016260928 0.98992
0.015216767 0.99534
0.012952065 0.9955
0.008880938 0.99868
0.006781239 0.99742
0.006611649 0.99772
0.007158364 0.99824
```

The results from running the Migration Script TF 2 version are: 
```
(train_loss, train_accuracy)
0.86170954 0.91768
0.23816809 0.9488
0.16750608 0.96032
0.12643586 0.96756
0.0971571 0.97194
0.07221356 0.9815
0.05393147 0.9875
0.04196974 0.98938
0.03490297 0.98796
0.029905643 0.98764
0.027600855 0.98814
0.026849583 0.9914
0.021233767 0.99068
0.015700594 0.99518
0.0131430775 0.99696
0.010740909 0.99608
0.0077435044 0.99426
0.0060562133 0.99568
0.006112121 0.99866
0.0034680425 0.99874
```

As we can see, they are significantly different. We ran this multiple times, the original version gave us consistent results whereas the compat version gave us varying results.

From the migration script, we get a warning:
```
31:21: INFO: Changing labels arg of tf.nn.softmax_cross_entropy_with_logits to tf.stop_gradient(labels). Please check this transformation.
```

**Describe the expected behavior**

The expected behavior was that the results between the Migration Script TF 2 version would yield similiar results to the original TF 1.x results.

**Standalone code to reproduce the issue**

You can reproduce this results running them on the following Colab Notebooks: 

TF 1.x Original Version: https://colab.research.google.com/drive/1OL_qPqdF9Hfhksqnri2BpJmbRc5LT84d?usp=sharing

TF 2 Migration Script version (with tf.stop_gradient): https://colab.research.google.com/drive/12sao4pWQ_x5tLLazux0LkfqeoqhnaCuL?usp=sharing

**Comment by Saduf2019** [comment]
I am able to replicate the issue reported, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/1c864f75d99bbf54f41ae4d81652b16b/untitled484.ipynb) and [2.x here](url)

**Comment by ymodak** [comment]
Can you please change your `keras` import to `tensorflow.keras` and try again.
```python
from tensorflow.keras.datasets import mnist
from tensorflow import keras
Output:
0.8170178 0.91674
0.23980403 0.94668
0.16789182 0.96192
0.13041496 0.971
0.09908046 0.97144
0.07447701 0.97742
0.059314076 0.98188
0.04840622 0.98608
0.04056797 0.9865
0.037097402 0.98898
0.031317815 0.99174
0.022105625 0.9934
0.019010812 0.99446
0.015984952 0.99136
0.012451912 0.99478
0.012425626 0.998
0.011714706 0.9949
0.009512669 0.99492
0.0075133997 0.9958
0.0081367865 0.9977
```

**Comment by tatianacv** [comment]
Hello, 

I have tried what was recommended but the output is still behaving unexpectedly. When I ran the code updated with your recommendations, sometimes I get an output that is close to the original code (TF 1.x) but other times the output is significantly different (see example of an output below).  So I don't believe that the change of changing keras import to tensorflow.keras worked to solve the unexpected behavior. 

```
0.9402422 0.9171
0.23724595 0.94824
0.15908638 0.9646
0.11688811 0.97304
0.08697565 0.97676
0.06634963 0.9817
0.052655857 0.9826
0.043853503 0.98538
0.041294258 0.9871
0.035308436 0.99142
0.024075007 0.9912
0.016586887 0.99578
0.013315758 0.9951
0.015305804 0.99648
0.010092826 0.99042
0.009263903 0.99702
0.0077375784 0.99706
0.0074062524 0.99856
0.0066001886 0.99648
0.0050791865 0.99838
```

**Comment by ymodak** [comment]
I got results on last epoch for multiple runs as :
```python
SEED = 123
os.environ['TF_DETERMINISTIC_OPS'] = '1'
os.environ['PYTHONHASHSEED']=str(SEED)
random.seed(SEED)
np.random.seed(SEED)
tf.random.set_seed(SEED)

#last epoch result on multiple runs:
0.004884758 0.9988
0.0038588147 0.99928
0.0040086186 0.99872
```
The inconsistency with precision `4` can be explained by gradient gating in `tf.compat.v1.train.AdamOptimizer`.
See https://github.com/NVIDIA/framework-determinism#gradient-gating to know more.

**Comment by tatianacv** [comment]
After adding what was said before,
```
SEED = 123
os.environ['TF_DETERMINISTIC_OPS'] = '1'
os.environ['PYTHONHASHSEED']=str(SEED)
random.seed(SEED)
np.random.seed(SEED)
tf.random.set_seed(SEED)
```
 I still do not get consistent results when comparing the results from TF 1.x code and TF 2.x code, as we would expect. The results I get are 
 
 ```
TF 2.x
 0.8727456 0.91456
0.23801954 0.94734
0.16402669 0.96284
0.12854795 0.9697
0.096198834 0.97468
0.07094173 0.98196
0.054136228 0.98686
0.043054197 0.99002
0.036161695 0.9898
0.032770615 0.98578
0.033984397 0.9837
0.033502463 0.99002
0.025235573 0.9934
0.015609697 0.99556
0.011150459 0.99752
0.009088196 0.9979
0.0058030323 0.99914
0.0038966627 0.99862
0.0031383259 0.99918
0.0020442707 0.99934
```

```
TF 1.x
0.8388736 0.91824
0.24215287 0.94656
0.17105241 0.96026
0.12800093 0.97012
0.095361866 0.97358
0.07280199 0.97786
0.05656019 0.98174
0.04541241 0.9866
0.03805694 0.98696
0.03796135 0.98726
0.033143476 0.99358
0.021085953 0.99454
0.015692161 0.9952
0.0122016715 0.99466
0.010179433 0.99418
0.009387239 0.99456
0.009418501 0.99698
0.007646892 0.99478
0.007249856 0.9975
0.009419065 0.99574
```

**Comment by saikumarchalla** [comment]
Was able to reproduce the issue and getting inconsistent results. Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/c4dcc9a7272e60c7766112e46f60ad31/untitled88.ipynb).Thanks!

**Comment by sachinprasadhs** [comment]
There will be floating point precision errors from Tensorflow 1.x to Tensorflow 2.x and it is not possible to get deterministic behavior from Tensorflow 1.x and Tensorflow 2.x results. The outputs will be almost close.
